From: Bastian Blank <waldi@debian.org>
Date: Sat, 5 Jul 2014 11:46:43 +0200
Subject: Display Debian package version in hypervisor log

During hypervisor boot, disable the banner and nicely display the xen
version as well as the Maintainer address from debian/control.

For this to work the DEB_VERSION and DEB_MAINTAINER variables needs to
be set by debian/rules.

Original patch by Bastian Blank <waldi@debian.org>
Modified by
Hans van Kranenburg <hans@knorrie.org>
Maximilian Engelhardt <maxi@daemonizer.de>
---
 xen/build.mk                 |  8 +++++---
 xen/common/kernel.c          |  8 ++++----
 xen/common/version.c         | 29 +++++++++++++++--------------
 xen/drivers/char/console.c   |  6 ++----
 xen/include/xen/compile.h.in | 10 ++++++----
 xen/include/xen/version.h    |  8 ++++----
 6 files changed, 36 insertions(+), 33 deletions(-)

diff --git a/xen/build.mk b/xen/build.mk
index 942ff38..872b529 100644
--- a/xen/build.mk
+++ b/xen/build.mk
@@ -17,7 +17,6 @@ targets += .banner
 quiet_cmd_compile.h = UPD     $@
 define cmd_compile.h
     if [ ! -r $@ -o -O $@ ]; then \
-	cat .banner; \
 	sed -e 's/@@date@@/$(XEN_BUILD_DATE)/g' \
 	    -e 's/@@time@@/$(XEN_BUILD_TIME)/g' \
 	    -e 's/@@whoami@@/$(XEN_WHOAMI)/g' \
@@ -28,13 +27,16 @@ define cmd_compile.h
 	    -e 's/@@subversion@@/$(XEN_SUBVERSION)/g' \
 	    -e 's/@@extraversion@@/$(XEN_EXTRAVERSION)/g' \
 	    -e 's!@@changeset@@!$(shell $(srctree)/tools/scmversion $(XEN_ROOT) || echo "unavailable")!g' \
+	    -e 's/@@system_distribution@@/$(shell lsb_release -is)/g' \
+	    -e 's/@@system_maintainer_domain@@/$(shell echo "${DEB_MAINTAINER}" | sed -ne 's,^.[^<]*<[^@>]*@\([^>]*\)>,\1,p')/g' \
+	    -e 's/@@system_maintainer_local@@/$(shell echo "${DEB_MAINTAINER}" | sed -ne 's,^.[^<]*<\([^@>]*\)@.*>,\1,p')/g' \
+	    -e 's/@@system_version@@/$(shell echo "${DEB_VERSION}")/g' \
 	    < $< > $(dot-target).tmp; \
-	sed -rf $(srctree)/tools/process-banner.sed < .banner >> $(dot-target).tmp; \
 	mv -f $(dot-target).tmp $@; \
     fi
 endef
 
-include/xen/compile.h: include/xen/compile.h.in .banner FORCE
+include/xen/compile.h: include/xen/compile.h.in FORCE
 	$(if $(filter-out FORCE,$?),$(Q)rm -fv $@)
 	$(call if_changed,compile.h)
 
diff --git a/xen/common/kernel.c b/xen/common/kernel.c
index fb45f81..38303ca 100644
--- a/xen/common/kernel.c
+++ b/xen/common/kernel.c
@@ -456,9 +456,9 @@ static int __init cf_check buildinfo_init(void)
 
     hypfs_add_dir(&buildinfo, &compileinfo, true);
     hypfs_string_set_reference(&compiler, xen_compiler());
-    hypfs_string_set_reference(&compile_by, xen_compile_by());
+    hypfs_string_set_reference(&compile_by, xen_compile_system_maintainer_local());
     hypfs_string_set_reference(&compile_date, xen_compile_date());
-    hypfs_string_set_reference(&compile_domain, xen_compile_domain());
+    hypfs_string_set_reference(&compile_domain, xen_compile_system_maintainer_domain());
     hypfs_add_leaf(&compileinfo, &compiler, true);
     hypfs_add_leaf(&compileinfo, &compile_by, true);
     hypfs_add_leaf(&compileinfo, &compile_date, true);
@@ -589,8 +589,8 @@ long do_xen_version(int cmd, XEN_GUEST_HANDLE_PARAM(void) arg)
 
         memset(&info, 0, sizeof(info));
         safe_strcpy(info.compiler,       deny ? xen_deny() : xen_compiler());
-        safe_strcpy(info.compile_by,     deny ? xen_deny() : xen_compile_by());
-        safe_strcpy(info.compile_domain, deny ? xen_deny() : xen_compile_domain());
+        safe_strcpy(info.compile_by,     deny ? xen_deny() : xen_compile_system_maintainer_local());
+        safe_strcpy(info.compile_domain, deny ? xen_deny() : xen_compile_system_maintainer_domain());
         safe_strcpy(info.compile_date,   deny ? xen_deny() : xen_compile_date());
         if ( copy_to_guest(arg, &info, 1) )
             return -EFAULT;
diff --git a/xen/common/version.c b/xen/common/version.c
index 80e7a9b..62d2d47 100644
--- a/xen/common/version.c
+++ b/xen/common/version.c
@@ -20,19 +20,24 @@ const char *xen_compile_time(void)
     return XEN_COMPILE_TIME;
 }
 
-const char *xen_compile_by(void)
+const char *xen_compile_system_distribution(void)
 {
-    return XEN_COMPILE_BY;
+    return XEN_COMPILE_SYSTEM_DISTRIBUTION;
 }
 
-const char *xen_compile_domain(void)
+const char *xen_compile_system_maintainer_local(void)
 {
-    return XEN_COMPILE_DOMAIN;
+    return XEN_COMPILE_SYSTEM_MAINTAINER_LOCAL;
 }
 
-const char *xen_compile_host(void)
+const char *xen_compile_system_maintainer_domain(void)
 {
-    return XEN_COMPILE_HOST;
+    return XEN_COMPILE_SYSTEM_MAINTAINER_DOMAIN;
+}
+
+const char *xen_compile_system_version(void)
+{
+    return XEN_COMPILE_SYSTEM_VERSION;
 }
 
 const char *xen_compiler(void)
@@ -60,11 +65,6 @@ const char *xen_changeset(void)
     return XEN_CHANGESET;
 }
 
-const char *xen_banner(void)
-{
-    return XEN_BANNER;
-}
-
 const char *xen_deny(void)
 {
     return "<denied>";
@@ -99,10 +99,11 @@ unsigned int __ro_after_init xen_build_id_len;
 
 void print_version(void)
 {
-    printk("Xen version %d.%d%s (%s@%s) (%s) %s %s\n",
+    printk("Xen version %d.%d%s (%s %s) (%s@%s) (%s) %s %s\n",
            xen_major_version(), xen_minor_version(), xen_extra_version(),
-           xen_compile_by(), xen_compile_domain(), xen_compiler(),
-           xen_build_info(), xen_compile_date());
+           xen_compile_system_distribution(), xen_compile_system_version(),
+           xen_compile_system_maintainer_local(), xen_compile_system_maintainer_domain(),
+           xen_compiler(), xen_build_info(), xen_compile_date());
 
     printk("Latest ChangeSet: %s\n", xen_changeset());
 
diff --git a/xen/drivers/char/console.c b/xen/drivers/char/console.c
index bcd6d26..837de7d 100644
--- a/xen/drivers/char/console.c
+++ b/xen/drivers/char/console.c
@@ -1134,10 +1134,8 @@ void __init console_init_preirq(void)
 
     conring_flush(flags);
 
-    /* HELLO WORLD --- start-of-day banner text. */
-    nrspin_lock(&console_lock);
-    __putstr(xen_banner());
-    nrspin_unlock(&console_lock);
+    /* HELLO WORLD --- start-of-day banner text.  The Debian version banner
+       is emitted by print_version() below. */
 
     /* Locate the buildid, if possible. */
     xen_build_init();
diff --git a/xen/include/xen/compile.h.in b/xen/include/xen/compile.h.in
index 9206341..e41b0cf 100644
--- a/xen/include/xen/compile.h.in
+++ b/xen/include/xen/compile.h.in
@@ -3,9 +3,10 @@
 
 #define XEN_COMPILE_DATE	"@@date@@"
 #define XEN_COMPILE_TIME	"@@time@@"
-#define XEN_COMPILE_BY		"@@whoami@@"
-#define XEN_COMPILE_DOMAIN	"@@domain@@"
-#define XEN_COMPILE_HOST	"@@hostname@@"
+#define XEN_COMPILE_SYSTEM_DISTRIBUTION		"@@system_distribution@@"
+#define XEN_COMPILE_SYSTEM_MAINTAINER_DOMAIN	"@@system_maintainer_domain@@"
+#define XEN_COMPILE_SYSTEM_MAINTAINER_LOCAL	"@@system_maintainer_local@@"
+#define XEN_COMPILE_SYSTEM_VERSION		"@@system_version@@"
 #define XEN_COMPILER		"@@compiler@@"
 
 #define XEN_VERSION		@@version@@
@@ -14,4 +15,5 @@
 #define XEN_EXTRAVERSION	"@@extraversion@@"
 
 #define XEN_CHANGESET		"@@changeset@@"
-#define XEN_BANNER		\
+
+#endif /* XEN_COMPILE_H */
diff --git a/xen/include/xen/version.h b/xen/include/xen/version.h
index 6f5d9c95..bc051e0 100644
--- a/xen/include/xen/version.h
+++ b/xen/include/xen/version.h
@@ -6,15 +6,15 @@
 
 const char *xen_compile_date(void);
 const char *xen_compile_time(void);
-const char *xen_compile_by(void);
-const char *xen_compile_domain(void);
-const char *xen_compile_host(void);
+const char *xen_compile_system_distribution(void);
+const char *xen_compile_system_maintainer_domain(void);
+const char *xen_compile_system_maintainer_local(void);
+const char *xen_compile_system_version(void);
 const char *xen_compiler(void);
 unsigned int xen_major_version(void);
 unsigned int xen_minor_version(void);
 const char *xen_extra_version(void);
 const char *xen_changeset(void);
-const char *xen_banner(void);
 const char *xen_deny(void);
 const char *xen_build_info(void);
 
